File: frontend/src/app/_models/scheduled-job.ts
﻿import {
    Workflow, WorkflowInfoList, WorkflowInfoTuple,
    Dataflow, DataflowInfoList,
    DataProtocol, DataProtocolInfoList,
    Protocol, Database,
} from '@app/_models';


export enum ScheduledJobStatus {
    WAITING_FOR_EXECUTION = 0,
    SUBMITED_FOR_EXECUTION = 1,
    RUNNING = 2,
    FINISHED_SUCCESSFUL = 3,
    FINISHED_ERROR = 4,
    FINISHED_MANUALLY = 5,
    UNKNOWN = 6,
}

export interface ScheduledJob {
    id?: number;
    // uuid?: string;
    title: string;
    description?: string;
    owner: string;
    owner_id: number;

    workflows?: Workflow[];
    dataflows?: Dataflow[];
    // data_protocols?: DataProtocol[];
    list_protocol_and_database?: ProtocolAndDatabaseTuple[];
    // database: string;

    execute_at: Date;
    created_at: Date;
    scheduled_at: Date;
    job_id: number;
    job_status: typeof ScheduledJobStatus;
}

export interface ScheduledJobInfo {
    id?: number;
    // uuid?: string;
    title: string;
    description?: string;
    owner?: string;
    owner_id?: number;

    workflows?: WorkflowInfoTuple[];  // WorkflowInfoList[];
    dataflows?: DataflowInfoList;
    // data_protocols?: DataProtocolInfoList;
    list_protocol_and_database?: [number, number][];
    // database?: string;

    execute_at: Date;
    created_at: Date;
    scheduled_at: Date;
    job_id: number;
    job_status: typeof  ScheduledJobStatus;
}

export interface ScheduledJobInfoList {
    data: ScheduledJobInfo[];
}

export interface ProtocolAndDatabaseTuple {
    protocol: Protocol;
    database: Database;
}

File: workflow-scheduler/app/app/schemas/scheduled_job.py
from typing import Optional, Tuple, List
from datetime import datetime
from enum import IntEnum

from pydantic import BaseModel


class ScheduledJobStatus(IntEnum):
    WAITING_FOR_EXECUTION = 0
    SUBMITTED_FOR_EXECUTION = 1
    RUNNING = 2
    FINISHED_SUCCESSFUL = 3
    FINISHED_ERROR = 4
    FINISHED_MANUALLY = 5
    UNKNOWN = 6


# Shared properties
class ScheduledJobBase(BaseModel):
    title: str = None
    description: Optional[str] = None
    owner: str = None
    owner_id: int = None

    workflows: Optional[List[Tuple[int, str, datetime]]] = []   # [workflow_id, workflow_type, workflow_execute_at]
    list_protocol_and_database: Optional[List[Tuple[int, int]]] = []   # [(protocol_id, database_id)]
    dataflow_path: Optional[str] = None

    # data_acquisition_protocol: Optional[Json] = None
    # data_acquisition_protocol_execute_at: Optional[datetime]
    # dataflow: Optional[Json] = None
    # dataflow_type: Optional[str] = None
    # dataflow_execute_at: Optional[datetime]
    # database: Optional[UUID]
    # service_bookings:
    execute_at: datetime = None
    created_at: datetime = None
    scheduled_at: datetime = None
    job_id: Optional[int] = None
    job_status: ScheduledJobStatus = ScheduledJobStatus(6)


# Properties to receive on item creation
class ScheduledJobCreate(ScheduledJobBase):
    pass


# Properties to receive on item update
class ScheduledJobUpdate(ScheduledJobBase):
    id: int
    job_id: int
    pass


# Properties shared by models stored in DB
class ScheduledJobInDBBase(ScheduledJobBase):
    id: int
    job_id: int
    class Config:
        orm_mode = True


# Properties to return to client
class ScheduledJob(ScheduledJobInDBBase):
    pass


# Properties properties stored in DB
class ScheduledJobInDB(ScheduledJobInDBBase):
    pass
